### PR TITLE
[Bug][FeedStorage][Hay] Hay dm bug fix

### LINKED ISSUE #3114: Hay DM mass goes negative
The hay DM mass can become negative and cause the simulation to crash. 

## Expected Behavior

The feed storage degradation functions should not allow the feed mass to be negative because we RuFaS should operate within the bounds of observable physical behavior (although I guess some forms of theoretical physics claim that negative mass is possible... who knew?). 

Once the dry matter degradations have been "capped" at reaching zero mass, the degradation functions (i.e. those in `process_degradations()` in `storage.py` should only be called if the crop's dry matter mass is greater than 0 to avoid division by zero errors. 

## Current Behavior

Hay DM mass can become negative in the current methods. This issue was raised through a FARM-ES evaluation that produced this error: 
`'<' not supported between instances of 'float' and 'complex'" in /app/RUFAS/biophysical/feed_storage/storage.py", line 837`. 

The line in question is the following:

`if `initial_nutrient_fraction` < fraction_of_nutrient_in_lost_dry_matter:`

After doing some more investigation with the farm that produced the errror, I found that when the error was produced, `initial_nutrient_fraction` was negative and `fraction_of_nutrient_in_lost_dry_matter` is a complex number 0 ( `0j`) and the crop that was fescue being storaged as a ProtectedTarp hay. 

Tracing the `initial_nutrient_fraction` back, I found that the stored crops dry matter mass is becoming negative.

## Possible Solution

 I started a branch off of `main` called `feed_storage_bug` that has some clumsy attempts to prevent the error but am now getting a division by zero error in `calculate_initial_dry_matter_loss_to_gas` in `hay.py`. I tried to prevent this error by skipping the degradation process all together if the crop's DM mass is 0 but was unsuccessful. 

## Steps to Reproduce

I am still working on a reproducible example that I can share and will follow up shortly. 

## Context (Environment)
This is simulation ending error that should not be feasible and will need to be pushed as a Patch update. 

## Target Deadline

Marking this as urgent so we should fix as soon as possible.

### PR DESCRIPTION
**Pull request by ew3361zh**
Fixes a bug where we were seeing negative dm amounts for `Hay` in storage.

### Context

Issue(s) closed by this pull request: closes #

### What

1. Caps dm loss to gas for hay so it can't exceed what's left in storage or go below zero).
2. Prevents 2 other divide by 0 errors that could be caused by capping those losses.

### Why

With the right combination of harvest schedule for hay and a large enough herd eating through the stored hay, we were seeing instances where the degradations calculations were creating projected losses greater than the amount left in storage which led to an error because that was not capped and negative dry matter makes no sense.

### How

In `Hay`.`calculate_dry_matter_loss_to_gas()`, it used to return this:
```python
return current_loss - processed_loss + additional_loss
```

That's now been bound between 0.0 and the available dry matter mass in storage, so instead we have:
```python
raw_loss = current_loss - processed_loss + additional_loss

return min(crop.dry_matter_mass, max(0.0, raw_loss))
```

With this fixed, 2 other related issues came up that were addressed here defensively:
- one (or more) of the harvests in the provided schedule input files yielded a 0.0 mass crop which meant it had 0.0% dm and 0.0 `initial_dry_matter`.
- Each of those was referenced as a denominator for a calculation in the `Storage` and `Hay` classes. Those have now been guarded against divide-by-zero errors.
- IMPORTANT TO NOTE FOR SMEs: please look at the `storage.py` `recalculate_nutrient_percentage()` divide-by-zero catch. I set the `dry_matter_loss_fraction` to 0.0 if the `initial_dry_matter` is <= 0.0. Just want to make sure that's correct since we have different returns for the if the fraction is 1.0 and I wasn't sure what a good substitute would be with 0.0 `initial_dry_matter`.

### Test plan

Have run this with the provided private test files and it was successful. Updating unit tests accordingly.

Need SME confirmation the science is correct in the fix.

### Output Changes

- N/A

#### Filter

```json

```

### CODE REVIEW

**Comment by ew3361zh** [comment]
redid with #3130 - issue with base branch comparison on this version.
